<h1>Salary salary MAE</h1>

In [104]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [105]:
salary = pd.read_csv("C:/ai_dataset/Salary_Data.csv")

In [106]:
y=salary.Salary.values
x=salary.drop('Salary',axis=1)

In [107]:
X = salary['Salary']

y = salary['YearsExperience']

In [108]:
X = X.to_numpy()
y = y.to_numpy()

In [109]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.25, random_state=0)

In [110]:
X_train = X_train.reshape(-1, 1)
X_test =  X_test.reshape(-1, 1)

In [111]:
from sklearn.linear_model import LinearRegression

ols_regressor = LinearRegression()
ols_regressor.fit(X_train,y_train)

LinearRegression()

In [112]:
y_pred = ols_regressor.predict(X_test)

<h2>Mean Absolute Error</h2>

<h5>기존 LinearRegression은 OLS 기반</h5>

In [113]:
from sklearn.metrics import mean_absolute_error

In [114]:
ols_y_prediction = ols_regressor.predict(X_test)

ols_mae = mean_absolute_error(y_test, ols_y_prediction)

print(ols_mae)

0.4085448443954782


In [115]:
print(ols_regressor.coef_, ols_regressor.intercept_)

[0.00010017] -2.4069397347271444


In [116]:
print(ols_regressor.predict([[10]]))

[-2.40593806]


<h3>Ordinary Least Squares</h3>

In [117]:
from statsmodels.formula.api import ols

ols_model = ols('Salary~YearsExperience', data=salary).fit()
ols_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.955
Method:                 Least Squares   F-statistic:                     622.5
Date:                Thu, 23 Mar 2023   Prob (F-statistic):           1.14e-20
Time:                        02:42:51   Log-Likelihood:                -301.44
No. Observations:                  30   AIC:                             606.9
Df Residuals:                      28   BIC:                             609.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept        2.579e+04   2273.053     11.347      0.000    2.11e+04    3.04e+04
YearsExperience  9449.9623    378.755     24.950      0.000    8674.119    1.02e+04
==============================================================================
Omnibus:                        2.140   Durbin-Watson:                   1.648
Prob(Omnibus):                  0.343   Jarque-Bera (JB):                1.569
Skew:                           0.363   Prob(JB):                        0.456
Kurtosis:                       2.147   Cond. No.                         13.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<h2>Decrease Mean Absolute Error</h2>

<h3>Data Scaling</h3>

In [118]:
from sklearn.preprocessing import StandardScaler

In [119]:
scaler = StandardScaler()

scaled_X_train = scaler.fit_transform(X_train)
scaled_X_test = scaler.transform(X_test)

<h3>Gradient Descent</h3>

In [123]:
from sklearn.linear_model import SGDRegressor

In [124]:
gd_regressor = SGDRegressor(loss='epsilon_insensitive', alpha=0.01, max_iter=1000)
gd_regressor.fit(scaled_X_train, y_train)

SGDRegressor(alpha=0.01, loss='epsilon_insensitive')

<h3>Result of MAE</h3>

In [125]:
gd_y_prediction = gd_regressor.predict(scaled_X_test)

gd_mae = mean_absolute_error(y_test, gd_y_prediction)
print(gd_mae)

0.5681035013982092


In [126]:
print(gd_regressor.coef_, gd_regressor.intercept_)

[2.25662613] [4.83466833]


In [127]:
print(gd_regressor.predict([[10]]))

[27.40092965]


<h3>Tune Hyperparameter</h3>

In [133]:
h_gd_regressor = SGDRegressor(loss='epsilon_insensitive', alpha=0.0001, max_iter=5000, learning_rate='adaptive')
h_gd_regressor.fit(scaled_X_train, y_train)

SGDRegressor(learning_rate='adaptive', loss='epsilon_insensitive',
             max_iter=5000)

<h3>Result of MAE</h3>

In [134]:
h_gd_y_prediction = h_gd_regressor.predict(scaled_X_test)

h_gd_mae = mean_absolute_error(y_test, h_gd_y_prediction)
print(h_gd_mae)

0.3552138447818593


<h3>Stochastic Gradient Descent</h3>

In [128]:
from sklearn.linear_model import SGDRegressor

In [129]:
sgd_regressor = SGDRegressor(loss='epsilon_insensitive', epsilon=0.1, max_iter=1000, tol=1e-3, random_state=0)
sgd_regressor.fit(scaled_X_train, y_train)

SGDRegressor(loss='epsilon_insensitive', random_state=0)

<h3>Result of MAE</h3>

In [130]:
sgd_y_prediction = sgd_regressor.predict(scaled_X_test)

sgd_mae = mean_absolute_error(y_test, sgd_y_prediction)

print(sgd_mae)

0.5568829145642519


In [131]:
print(sgd_regressor.coef_, sgd_regressor.intercept_)

[2.26634043] [4.83890401]


In [132]:
print(sgd_regressor.predict([[10]]))

[27.50230831]


<h3>Tune Hyperparameter</h3>

In [135]:
h_sgd_regressor = SGDRegressor(loss='epsilon_insensitive', alpha=0.0001, max_iter=5000, learning_rate='adaptive', penalty='elasticnet', l1_ratio=0.1)
h_sgd_regressor.fit(scaled_X_train, y_train)

SGDRegressor(l1_ratio=0.1, learning_rate='adaptive', loss='epsilon_insensitive',
             max_iter=5000, penalty='elasticnet')

<h3>Result of MAE</h3>

In [137]:
h_sgd_y_prediction = h_sgd_regressor.predict(scaled_X_test)

h_sgd_mae = mean_absolute_error(y_test, h_sgd_y_prediction)
print(h_sgd_mae)

0.3652134643724427


<h2>Result</h2>

<hr>
<h3>MAE of Ordinary Least Square</h3>
<li>0.4085448443954782</li>
<hr>
<h3>MAE of Gradient Descent</h3>
<li>after data scaling : 0.5681035013982092</li>
<li>after tune hyperparameters : 0.3552138447818593</li>
<hr>
<h3>MAE of Stochastic Gradient Descent</h3>
<li>after data scaling : 0.5568829145642519</li>
<li>after tune hyperparameters : 0.3652134643724427</li>